File: rules-interpreter/src/test/java/MathTest.java
import com.github.iamrenny.ruleflow.Workflow;
import com.github.iamrenny.ruleflow.vo.WorkflowResult;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Map;
import java.util.Set;

class MathTest {

    @Test
    public void given_math_expression_when_multiplication_must_be_resolved_first() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * z = 15 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(Map.of("x", 15, "y", 22, "z", 0))
        );
    }

    @Test
    public void given_math_expression_with_constant_multiplication_must_be_resolved_first() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * 0 = 15 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(Map.of("x", 15, "y", 22))
        );
    }

    @Test
    public void given_math_expression_with_parenthesis_inner_expression_must_be_resolved_first() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' (x + y) * z = 80 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(Map.of("x", 15, "y", 25, "z", 2))
        );
    }

    @Test
    public void given_math_expression_with_multiplication_must_be_resolved_first() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * z = 65 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(Map.of("x", 15, "y", 25, "z", 2))
        );
    }

    @Test
    public void given_complex_math_expression_must_be_resolved_correctly() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y - 3 * z = -47.54 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(Map.of("x", 3.43, "y", 0, "z", 16.99))
        );
    }

    @Test
    public void given_division_expression_must_be_resolved_correctly() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x / y = 2 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(Map.of("x", 4, "y", 2))
        );
    }

    @Test
    public void given_division_by_zero_must_result_in_warning() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x / y = 2 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "default", "default", "allow", Set.of()),
            ruleEngine.evaluate(Map.of("x", 4, "y", 0))
        );
    }

    @Test
    public void given_nonexistent_field_in_expression_must_return_warning() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x / y = 2 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "default", "default", "allow", Set.of("y field cannot be found")),
            ruleEngine.evaluate(Map.of("x", 4))
        );
    }

    @Test
    public void given_modulo_expression_must_be_resolved_correctly() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x % y = 9 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(Map.of("x", 99, "y", 10))
        );
    }

    @Test
    public void given_modulo_expression_when_result_is_not_equal_must_return_default() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x % y = 9 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "default", "default", "allow"),
            ruleEngine.evaluate(Map.of("x", 99, "y", 100))
        );
    }

    @Test
    public void given_modulo_expression_with_nonexistent_right_value_must_return_warning() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x % y = 9 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "default", "default", "allow", Set.of("y field cannot be found")),
            ruleEngine.evaluate(Map.of("x", 9))
        );
    }

    @Test
    public void given_modulo_expression_with_nonexistent_left_value_must_return_warning() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x % y = 9 return block
                default allow
            end
        """;
        Workflow ruleEngine = new Workflow(workflow);
        Assertions.assertEquals(
            new WorkflowResult("test", "default", "default", "allow", Set.of("x field cannot be found")),
            ruleEngine.evaluate(Map.of("y", 10))
        );
    }
}

File: rules-interpreter/src/test/java/MathTest.kt
import com.github.iamrenny.ruleflow.Workflow
import com.github.iamrenny.ruleflow.vo.WorkflowResult
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class MathTest {
    @Test
    fun `given a math expression when operating a multiplication must be resolved first`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * z  = 15 return block
                default allow
            end
        """

        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 15,
                    "y" to 22,
                    "z" to 0
                    )
                )
        )
    }

    @Test
    fun `given a math expression when operating multiplication with constant must be resolved first`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * 0  = 15 return block
                default allow
            end
        """

        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 15,
                    "y" to 22
                )
            )
        )
    }


    @Test
    fun `given a math expression when using parenthesis then inner expression must be resolved first`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' (x + y) * z  = 80 return block
                default allow
            end
        """

        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 15,
                    "y" to 25,
                    "z" to 2
                )
            )
        )
    }

    @Test
    fun `given a math expression when multiplication expression must be resolved first`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y * z  = 65 return block
                default allow
            end
        """

        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 15,
                    "y" to 25,
                    "z" to 2
                )
            )
        )
    }

    @Test
    fun `given a complex math expression when operating must be resolved ok`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x + y - 3 * z  = -47.54 return block
                default allow
            end
        """
        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 3.43,
                    "y" to 0,
                    "z" to 16.99
                )
            )
        )
    }

    @Test
    fun `given a division  expression when operating must be resolved ok`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x / y = 2 return block
                default allow
            end
        """
        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 4,
                    "y" to 2
                )
            )
        )
    }

    @Test
    fun `given a division by zero expression when operating must result in a warning`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x / y = 2 return block
                default allow
            end
        """
        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "default", "default", "allow", warnings = setOf()),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 4,
                    "y" to 0
                )
            )
        )
    }

    @Test
    fun `given an expression with a non existant field when operating must return a warning`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x / y = 2 return block
                default allow
            end
        """
        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "default", "default", "allow", warnings = setOf("y field cannot be found")),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 4
                )
            )
        )
    }

    @Test
    fun `given a modulo  expression when operating must be resolved ok`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x % y = 9 return block
                default allow
            end
        """
        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "dummy", "item_a", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 99,
                    "y" to 10
                )
            )
        )
    }

    @Test
    fun `given a modulo expression when operating must be not equal to expected result`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x % y = 9 return block
                default allow
            end
        """
        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "default", "default", "allow"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 99,
                    "y" to 100
                )
            )
        )
    }

    @Test
    fun `given a modulo expression when operating with non existing right value must return warning`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x % y = 9 return block
                default allow
            end
        """
        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "default", "default", "allow", warnings =  setOf("y field cannot be found")),
            ruleEngine.evaluate(
                mapOf(
                    "x" to 9
                )
            )
        )
    }

    @Test
    fun `given a modulo expression when operating with non existing left value must return warning`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'item_a' x % y = 9 return block
                default allow
            end
        """
        val ruleEngine = com.github.iamrenny.ruleflow.Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult("test", "default", "default", "allow", warnings =  setOf("x field cannot be found")),
            ruleEngine.evaluate(
                mapOf(
                    "y" to 10
                )
            )
        )
    }
}